pkg/operator/etcdmemberlistercontroller/etcdmemberlistercontroller.go

```diff
@@ -0,0 +1,195 @@
+package etcdmemberlistercontroller
+
+import (
+	"context"
+	"reflect"
+	"sync"
+
+	errorsutil "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/klog/v2"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/cluster-etcd-operator/pkg/dnshelpers"
+	"github.com/openshift/cluster-etcd-operator/pkg/etcdcli"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+
+	"go.etcd.io/etcd/api/v3/etcdserverpb"
+)
+
+// Listener is an interface to use to notify interested parties of a change.
+type Listener interface {
+	// Enqueue should be called to update the listener.
+	Update()
+}
+
+// Notifier is an interface to add listeners for updates upon a change.
+type Notifier interface {
+	// AddListener is adds a listener to be notified.
+	AddListener(listener Listener)
+}
+
+// EtcdMemberLister is an interface to notify listeners upon etcd members list changes.
+type EtcdMemberLister interface {
+	Notifier
+	GetMembers() []etcdserverpb.Member
+}
+
+// membersCache uses a mutex to protect etcd membership map
+type membersCache struct {
+	mux   sync.RWMutex
+	cache map[string]etcdserverpb.Member
+}
+
+// newMembersCache returns an initialized and empty membersCache
+func newMembersCache() *membersCache {
+	return &membersCache{
+		mux:   sync.RWMutex{},
+		cache: make(map[string]etcdserverpb.Member),
+	}
+}
+
+// EtcdMemberListerController keeps etcd members cached, and updates the cache when the membership changes
+type EtcdMemberListerController struct {
+	membersCache            *membersCache
+	listeners               []Listener
+	lastConsecutiveFailures int
+	operatorClient          v1helpers.OperatorClient
+	etcdClient              etcdcli.EtcdClient
+}
+
+func NewEtcdMemberListerController(operatorClient v1helpers.OperatorClient,
+	etcdClient etcdcli.EtcdClient,
+	eventRecorder events.Recorder,
+) *EtcdMemberListerController {
+	c := &EtcdMemberListerController{
+		operatorClient:          operatorClient,
+		etcdClient:              etcdClient,
+		membersCache:            newMembersCache(),
+		listeners:               []Listener{},
+		lastConsecutiveFailures: 0,
+	}
+	//return factory.New().ResyncEvery(30*time.Second).WithSync(c.sync).ToController("EtcdMemberListerController", eventRecorder.WithComponentSuffix("etcd-member-lister-controller"))
```

Comment:
I do not know how to replace this to still get a controller while returning the `*EtcdMemberListerController` :/

Reply:
You have two options
* see how the `NewEnvVarController` does this, it doesn't use the factory, it merely implements `Run`
* compose the two interfaces

The latter looks clearer to me, think of something like that:

```
type ListerController interface {
	factory.Controller
	EtcdMemberLister
}
```

Then just return the ListerController from New*

Reply:
this is a great solution, thx so much .. i was stuck debugging the test cases :D

Reply:
no it is not working :/ 

```
Running `gofmt -s -l` on 97 file(s).
go vet -mod=vendor ./...
# github.com/openshift/cluster-etcd-operator/pkg/operator/etcdmemberlistercontroller
pkg/operator/etcdmemberlistercontroller/etcdmemberlistercontroller.go:79:9: cannot use factory.New().ResyncEvery(30 * time.Second).WithSync(c.sync).ToController("EtcdMemberListerController", eventRecorder.WithComponentSuffix("etcd-member-lister-controller")) (value of type factory.Controller) as type ListerController in return statement:
        factory.Controller does not implement ListerController (missing AddListener method)
# github.com/openshift/cluster-etcd-operator/pkg/operator/etcdmemberlistercontroller
vet: pkg/operator/etcdmemberlistercontroller/etcdmemberlistercontroller.go:79:9: cannot use factory.New().ResyncEvery(30 * time.Second).WithSync(c.sync).ToController("EtcdMemberListerController", eventRecorder.WithComponentSuffix("etcd-member-lister-controller")) (value of type factory.Controller) as ListerController value in return statement: factory.Controller does not implement ListerController (missing method AddListener)
make: *** [verify-govet] Error 2

```

Reply:
why don't you implement what it says then? :roll_eyes: 
> (missing AddListener method)